# Conociendo el problema

Se trata de un algoritmo capaz de analizar las caracteristicas de una propiedad -como comodidades, tama;o, ocupacion del espacioen un periodo determinado- y sugerir al anfitrion un precio a cobrar por tarifas diarias que garantice ganancias en momentos de alta demanda

## Descripcion de los datos
Cuando trabajamos con cualquier conjunto de datos, necesitamos saber qué información nos aportan esos datos, porque solo así podremos estudiarlos y analizarlos para desarrollar una solución de análisis y procesamiento de datos.

En este curso trabajaremos con el conjunto de datos presentes en el archivo datos_hosting.json y, para avanzar en nuestros estudios sobre los datos que proporciona este archivo, entenderemos qué información trae cada columna.

* evaluacion_general: se refiere a la puntuación media otorgada para evaluar el alojamiento en la propiedad.
* experiencia_local: describe las experiencias ofrecidas durante su estancia en la propiedad.
* max_hospedes: informa el número máximo de invitados que permite la ubicación.
* descripcion_local: describe la propiedad.
* descripcion_vecindad: describe el vecindario alrededor de la propiedad.
* cantidad_baños: informa el número de baños disponibles.
* cantidad_cuartos: informa el número de habitaciones disponibles.
* cantidad_camas: informa el número de camas disponibles.
* modelo_cama: informa el modelo de cama ofrecido.
* comodidades: informa las comodidades que ofrece la propiedad.
* cuota_deposito: informa la tarifa mínima de depósito para la seguridad del hosting.
* cuota_limpieza: informa el cargo cobrado por el servicio de limpieza.
* precio: se refiere al precio base a cobrar por la estancia diaria en la propiedad.

## Entendiendo el problema

In [34]:
import pandas as pd

In [35]:
archivo = 'datos_hosting.json'
datos = pd.read_json(archivo)

In [36]:
datos.head()

info_inmuebles
0  {'evaluacion_general': '10.0', 'experiencia_lo...
1  {'evaluacion_general': '10.0', 'experiencia_lo...
2  {'evaluacion_general': '10.0', 'experiencia_lo...
3  {'evaluacion_general': '10.0', 'experiencia_lo...
4  {'evaluacion_general': '10.0', 'experiencia_lo...

In [37]:
datos = pd.json_normalize(datos['info_inmuebles'])

In [38]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   evaluacion_general    70 non-null     object
 1   experiencia_local     70 non-null     object
 2   max_hospedes          70 non-null     object
 3   descripcion_local     70 non-null     object
 4   descripcion_vecindad  70 non-null     object
 5   cantidad_baños        70 non-null     object
 6   cantidad_cuartos      70 non-null     object
 7   cantidad_camas        70 non-null     object
 8   modelo_cama           70 non-null     object
 9   comodidades           70 non-null     object
 10  cuota_deposito        70 non-null     object
 11  cuota_limpieza        70 non-null     object
 12  precio                70 non-null     object
dtypes: object(13)
memory usage: 7.2+ KB


In [39]:
datos.head()

evaluacion_general experiencia_local max_hospedes  \
0               10.0                --            1   
1               10.0                --           10   
2               10.0                --           11   
3               10.0                --           12   
4               10.0                --           14   

                                   descripcion_local  \
0  [This clean and comfortable one bedroom sits r...   
1  [Welcome to the heart of the 'Ballard Brewery ...   
2  [New modern house built in 2013.  Spectacular ...   
3  [Our NW style home is 3200+ sq ft with 3 level...   
4  [Perfect for groups. 2 bedrooms, full bathroom...   

                                descripcion_vecindad  \
0  [Lower Queen Anne is near the Seattle Center (...   
1  [--, Capital Hill is the heart of Seattle, bor...   
2  [Upper Queen Anne is a charming neighborhood f...   
3  [The Views from our top floor!  Wallingford ha...   
4  [Safeway grocery store within walking distance...   

                                      cantidad_baños  \
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                  [2, 3, 2, 3, 3, 3, 2, 1, 2, 2, 2]   
2                                                [4]   
3                           [3, 3, 3, 3, 3, 3, 3, 3]   
4                                             [2, 3]   

                                    cantidad_cuartos  \
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                  [3, 4, 2, 3, 3, 3, 3, 3, 3, 4, 3]   
2                                                [5]   
3                           [6, 6, 5, 5, 5, 5, 4, 4]   
4                                             [2, 6]   

                                      cantidad_camas  \
0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...   
1                  [5, 6, 8, 3, 3, 5, 4, 5, 6, 7, 4]   
2                                                [7]   
3                           [6, 6, 7, 8, 7, 7, 6, 6]   
4                                             [3, 9]   

                                         modelo_cama  \
0  [Real Bed, Futon, Futon, Pull-out Sofa, Real B...   
1  [Real Bed, Real Bed, Real Bed, Real Bed, Real ...   
2                                         [Real Bed]   
3  [Real Bed, Real Bed, Real Bed, Real Bed, Real ...   
4                               [Real Bed, Real Bed]   

                                         comodidades  \
0  [{Internet,"Wireless Internet",Kitchen,"Free P...   
1  [{TV,Internet,"Wireless Internet",Kitchen,"Fre...   
2  [{TV,"Cable TV",Internet,"Wireless Internet","...   
3  [{Internet,"Wireless Internet",Kitchen,"Free P...   
4  [{TV,Internet,"Wireless Internet",Kitchen,"Fre...   

                                      cuota_deposito  \
0  [$0, $0, $0, $0, $0, $350.00, $350.00, $350.00...   
1  [$500.00, $300.00, $0, $300.00, $300.00, $360....   
2                                        [$1,000.00]   
3  [$500.00, $500.00, $500.00, $500.00, $500.00, ...   
4                               [$300.00, $2,000.00]   

                                      cuota_limpieza  \
0  [$0, $0, $0, $20.00, $15.00, $28.00, $35.00, $...   
1  [$125.00, $100.00, $85.00, $110.00, $110.00, $...   
2                                          [$300.00]   
3  [$225.00, $300.00, $250.00, $250.00, $250.00, ...   
4                                  [$40.00, $150.00]   

                                              precio  
0  [$110.00, $45.00, $55.00, $52.00, $85.00, $50....  
1  [$350.00, $300.00, $425.00, $300.00, $285.00, ...  
2                                          [$975.00]  
3  [$490.00, $550.00, $350.00, $350.00, $350.00, ...  
4                                 [$200.00, $545.00]

## Datos numericos

In [40]:
columnas = list(datos.columns)
columnas

['evaluacion_general',
 'experiencia_local',
 'max_hospedes',
 'descripcion_local',
 'descripcion_vecindad',
 'cantidad_baños',
 'cantidad_cuartos',
 'cantidad_camas',
 'modelo_cama',
 'comodidades',
 'cuota_deposito',
 'cuota_limpieza',
 'precio']

In [41]:
datos = datos.explode(columnas[3:])
datos

evaluacion_general experiencia_local max_hospedes  \
0                10.0                --            1   
0                10.0                --            1   
0                10.0                --            1   
0                10.0                --            1   
0                10.0                --            1   
..                ...               ...          ...   
68                nan                --            8   
68                nan                --            8   
68                nan                --            8   
69                nan                --            9   
69                nan                --            9   

                                    descripcion_local  \
0   This clean and comfortable one bedroom sits ri...   
0   Our century old Upper Queen Anne house is loca...   
0   Cozy room in two-bedroom apartment along the l...   
0   Very lovely and cozy room for one. Convenientl...   
0   The “Studio at Mibbett Hollow' is in a Beautif...   
..                                                ...   
68  Beautiful craftsman home in the historic Wedgw...   
68  Located in a very easily accessible area of Se...   
68  This home is fully furnished and available wee...   
69  This business-themed modern home features:  *H...   
69  This welcoming home is in the quiet residentia...   

                                 descripcion_vecindad cantidad_baños  \
0   Lower Queen Anne is near the Seattle Center (s...              1   
0   Upper Queen Anne is a really pleasant, unique ...              1   
0   The convenience of being in Seattle but on the...              1   
0   Ballard is lovely, vibrant and one of the most...              1   
0                                                  --              1   
..                                                ...            ...   
68                                                 --              3   
68  Quiet, dead end street near I-5. The proximity...              2   
68                                                 --              1   
69  Your hosts made Madison Valley their home when...              2   
69  Despite the quiet, leafy streets of this resid...              1   

   cantidad_cuartos cantidad_camas    modelo_cama  \
0                 1              1       Real Bed   
0                 1              1          Futon   
0                 1              1          Futon   
0                 1              1  Pull-out Sofa   
0                 1              1       Real Bed   
..              ...            ...            ...   
68                4              5       Real Bed   
68                4              4       Real Bed   
68                3              4       Real Bed   
69                3              6       Real Bed   
69                4              6       Real Bed   

                                          comodidades cuota_deposito  \
0   {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
0   {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
0   {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
0   {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
0   {"Wireless Internet",Kitchen,"Free Parking on ...             $0   
..                                                ...            ...   
68  {TV,"Cable TV",Internet,"Wireless Internet","A...      $1,000.00   
68  {TV,"Cable TV",Internet,"Wireless Internet",Ki...             $0   
68  {TV,"Cable TV",Internet,"Wireless Internet","A...             $0   
69  {TV,"Cable TV",Internet,"Wireless Internet","A...      $1,000.00   
69  {TV,Internet,"Wireless Internet",Kitchen,"Free...        $500.00   

   cuota_limpieza   precio  
0              $0  $110.00  
0              $0   $45.00  
0              $0   $55.00  
0          $20.00   $52.00  
0          $15.00   $85.00  
..            ...      ...  
68        $178.00  $299.00  
68         $99.00  $199.00  
68             $0

In [42]:
#reseteamos los indices
datos.reset_index(inplace = True, drop = True)
datos.head()

evaluacion_general experiencia_local max_hospedes  \
0               10.0                --            1   
1               10.0                --            1   
2               10.0                --            1   
3               10.0                --            1   
4               10.0                --            1   

                                   descripcion_local  \
0  This clean and comfortable one bedroom sits ri...   
1  Our century old Upper Queen Anne house is loca...   
2  Cozy room in two-bedroom apartment along the l...   
3  Very lovely and cozy room for one. Convenientl...   
4  The “Studio at Mibbett Hollow' is in a Beautif...   

                                descripcion_vecindad cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...              1   
1  Upper Queen Anne is a really pleasant, unique ...              1   
2  The convenience of being in Seattle but on the...              1   
3  Ballard is lovely, vibrant and one of the most...              1   
4                                                 --              1   

  cantidad_cuartos cantidad_camas    modelo_cama  \
0                1              1       Real Bed   
1                1              1          Futon   
2                1              1          Futon   
3                1              1  Pull-out Sofa   
4                1              1       Real Bed   

                                         comodidades cuota_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
1  {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
2  {TV,Internet,"Wireless Internet",Kitchen,"Free...             $0   
3  {Internet,"Wireless Internet",Kitchen,"Free Pa...             $0   
4  {"Wireless Internet",Kitchen,"Free Parking on ...             $0   

  cuota_limpieza   precio  
0             $0  $110.00  
1             $0   $45.00  
2             $0   $55.00  
3         $20.00   $52.00  
4         $15.00   $85.00

In [43]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   evaluacion_general    3818 non-null   object
 1   experiencia_local     3818 non-null   object
 2   max_hospedes          3818 non-null   object
 3   descripcion_local     3818 non-null   object
 4   descripcion_vecindad  3818 non-null   object
 5   cantidad_baños        3818 non-null   object
 6   cantidad_cuartos      3818 non-null   object
 7   cantidad_camas        3818 non-null   object
 8   modelo_cama           3818 non-null   object
 9   comodidades           3818 non-null   object
 10  cuota_deposito        3818 non-null   object
 11  cuota_limpieza        3818 non-null   object
 12  precio                3818 non-null   object
dtypes: object(13)
memory usage: 387.9+ KB


### Convirtiendo a Datos numericos

In [44]:
import numpy as np

In [45]:
col_numericas = ['cantidad_baños','cantidad_cuartos', 'cantidad_camas']

In [46]:
datos['evaluacion_general'] = datos['evaluacion_general'].astype(np.float64)
datos['max_hospedes'] = datos['max_hospedes'].astype(np.int64)
datos[col_numericas] = datos[col_numericas].astype(np.int64)


In [47]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   evaluacion_general    3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descripcion_local     3818 non-null   object 
 4   descripcion_vecindad  3818 non-null   object 
 5   cantidad_baños        3818 non-null   int64  
 6   cantidad_cuartos      3818 non-null   int64  
 7   cantidad_camas        3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  cuota_deposito        3818 non-null   object 
 11  cuota_limpieza        3818 non-null   object 
 12  precio                3818 non-null   object 
dtypes: float64(1), int64(4), object(8)
memory usage: 387.9+ KB


In [48]:
col_dolar = ['cuota_deposito', 'cuota_limpieza', 'precio']

In [49]:
datos[col_dolar] = datos[col_dolar].map(lambda x: x.replace('$', '').replace('.', '').replace(',','.').strip()).astype(np.float64) #strip elimina espacios en blanco

In [50]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
1                10.0                --             1   
2                10.0                --             1   
3                10.0                --             1   
4                10.0                --             1   

                                   descripcion_local  \
0  This clean and comfortable one bedroom sits ri...   
1  Our century old Upper Queen Anne house is loca...   
2  Cozy room in two-bedroom apartment along the l...   
3  Very lovely and cozy room for one. Convenientl...   
4  The “Studio at Mibbett Hollow' is in a Beautif...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
1  Upper Queen Anne is a really pleasant, unique ...               1   
2  The convenience of being in Seattle but on the...               1   
3  Ballard is lovely, vibrant and one of the most...               1   
4                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
1                 1               1          Futon   
2                 1               1          Futon   
3                 1               1  Pull-out Sofa   
4                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
1  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
2  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
3  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
4  {"Wireless Internet",Kitchen,"Free Parking on ...             0.0   

   cuota_limpieza   precio  
0             0.0  11000.0  
1             0.0   4500.0  
2             0.0   5500.0  
3          2000.0   5200.0  
4          1500.0   8500.0

In [51]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   evaluacion_general    3162 non-null   float64
 1   experiencia_local     3818 non-null   object 
 2   max_hospedes          3818 non-null   int64  
 3   descripcion_local     3818 non-null   object 
 4   descripcion_vecindad  3818 non-null   object 
 5   cantidad_baños        3818 non-null   int64  
 6   cantidad_cuartos      3818 non-null   int64  
 7   cantidad_camas        3818 non-null   int64  
 8   modelo_cama           3818 non-null   object 
 9   comodidades           3818 non-null   object 
 10  cuota_deposito        3818 non-null   float64
 11  cuota_limpieza        3818 non-null   float64
 12  precio                3818 non-null   float64
dtypes: float64(4), int64(4), object(5)
memory usage: 387.9+ KB


## Datos de texto

In [52]:
datos['descripcion_local'] = datos['descripcion_local'].str.lower() #Primero llevamoms a minusculas

#### REGEX

In [53]:
datos['descripcion_local'][10]

"built, run and supported by seattle tech and start up veterans, grokhome's focus is to create a supportive environment for smart people working on interesting projects, start ups and more. this listing is an upper bunk, in a 4-person shared room. ** this is an all-female room. *note: this fall, there will be major renovations happening on one kitchen and bathroom at a time. there will always be two other working kitchens and two working bathrooms in the house. we'll work to minimize the impact these renovations have on your stay. **this listing is only available to those working in the tech/science space. live in a hacker house, and immerse yourself in the seattle tech scene. you can expect to be surrounded by smart people solving big problems or working on something fun. we have frequent demo nights, and love when our guests share something they are passionate about. if you're new to the city, our deep ties to the seattle tech scene can help you get involved. expand your network, dev

In [54]:
datos['descripcion_local'] = datos['descripcion_local'].str.replace('[^a-zA-Z0-9\-\' ]',' ', regex=True)
# ^ Significa negacion, osea que va a reemplazar todo lo que no corresponda a lo que tenga despues
#a-zA-Z significa que tengo letras de la a a la z, de A a Z y los numeros
# '\-' significa que no sustituya los guiones
# '\'' tammpoco quiero que reemplaze las comillas

<>:1: SyntaxWarning: invalid escape sequence '\-'
<>:1: SyntaxWarning: invalid escape sequence '\-'
C:\Users\caray\AppData\Local\Temp\ipykernel_20204\3637815317.py:1: SyntaxWarning: invalid escape sequence '\-'
  datos['descripcion_local'] = datos['descripcion_local'].str.replace('[^a-zA-Z0-9\-\' ]',' ', regex=True)


In [55]:
datos['descripcion_local'][10]

"built  run and supported by seattle tech and start up veterans  grokhome's focus is to create a supportive environment for smart people working on interesting projects  start ups and more  this listing is an upper bunk  in a 4-person shared room     this is an all-female room   note  this fall  there will be major renovations happening on one kitchen and bathroom at a time  there will always be two other working kitchens and two working bathrooms in the house  we'll work to minimize the impact these renovations have on your stay    this listing is only available to those working in the tech science space  live in a hacker house  and immerse yourself in the seattle tech scene  you can expect to be surrounded by smart people solving big problems or working on something fun  we have frequent demo nights  and love when our guests share something they are passionate about  if you're new to the city  our deep ties to the seattle tech scene can help you get involved  expand your network  dev

In [56]:
datos['descripcion_local'] = datos['descripcion_local'].str.replace('(?<!\w)-(?!\w)',' ', regex=True)
# (?<!\w), esto significa que buscara si tiene algo acompa;ando al '-', si es que no hay caracter, se reemplza. Esto sirve para separar palabras como 
# one - head

<>:1: SyntaxWarning: invalid escape sequence '\w'
<>:1: SyntaxWarning: invalid escape sequence '\w'
C:\Users\caray\AppData\Local\Temp\ipykernel_20204\3505384210.py:1: SyntaxWarning: invalid escape sequence '\w'
  datos['descripcion_local'] = datos['descripcion_local'].str.replace('(?<!\w)-(?!\w)',' ', regex=True)


In [57]:
datos['descripcion_local'].head()

0    this clean and comfortable one bedroom sits ri...
1    our century old upper queen anne house is loca...
2    cozy room in two-bedroom apartment along the l...
3    very lovely and cozy room for one  convenientl...
4    the  studio at mibbett hollow' is in a beautif...
Name: descripcion_local, dtype: object

In [ ]:
datos['descripcion_local'] = datos['descripcion_local'].str.split() #Se separan las palabras

In [59]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
1                10.0                --             1   
2                10.0                --             1   
3                10.0                --             1   
4                10.0                --             1   

                                   descripcion_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
1  [our, century, old, upper, queen, anne, house,...   
2  [cozy, room, in, two-bedroom, apartment, along...   
3  [very, lovely, and, cozy, room, for, one, conv...   
4  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
1  Upper Queen Anne is a really pleasant, unique ...               1   
2  The convenience of being in Seattle but on the...               1   
3  Ballard is lovely, vibrant and one of the most...               1   
4                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
1                 1               1          Futon   
2                 1               1          Futon   
3                 1               1  Pull-out Sofa   
4                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
1  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
2  {TV,Internet,"Wireless Internet",Kitchen,"Free...             0.0   
3  {Internet,"Wireless Internet",Kitchen,"Free Pa...             0.0   
4  {"Wireless Internet",Kitchen,"Free Parking on ...             0.0   

   cuota_limpieza   precio  
0             0.0  11000.0  
1             0.0   4500.0  
2             0.0   5500.0  
3          2000.0   5200.0  
4          1500.0   8500.0

In [60]:
datos['comodidades'] = datos['comodidades'].str.replace('\{|}|\"','',regex=True)
# Esto busca las llaves {}, la '|' significa que debe buuscar otra cosa mas, en este caso las "

<>:1: SyntaxWarning: invalid escape sequence '\{'
<>:1: SyntaxWarning: invalid escape sequence '\{'
C:\Users\caray\AppData\Local\Temp\ipykernel_20204\1708022596.py:1: SyntaxWarning: invalid escape sequence '\{'
  datos['comodidades'] = datos['comodidades'].str.replace('\{|}|\"','',regex=True)


In [61]:
datos['comodidades']

0       Internet,Wireless Internet,Kitchen,Free Parkin...
1       TV,Internet,Wireless Internet,Kitchen,Free Par...
2       TV,Internet,Wireless Internet,Kitchen,Free Par...
3       Internet,Wireless Internet,Kitchen,Free Parkin...
4       Wireless Internet,Kitchen,Free Parking on Prem...
                              ...                        
3813    TV,Cable TV,Internet,Wireless Internet,Air Con...
3814    TV,Cable TV,Internet,Wireless Internet,Kitchen...
3815    TV,Cable TV,Internet,Wireless Internet,Air Con...
3816    TV,Cable TV,Internet,Wireless Internet,Air Con...
3817    TV,Internet,Wireless Internet,Kitchen,Free Par...
Name: comodidades, Length: 3818, dtype: object

In [62]:
datos['comodidades'] = datos['comodidades'].str.split(',')

In [63]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
1                10.0                --             1   
2                10.0                --             1   
3                10.0                --             1   
4                10.0                --             1   

                                   descripcion_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
1  [our, century, old, upper, queen, anne, house,...   
2  [cozy, room, in, two-bedroom, apartment, along...   
3  [very, lovely, and, cozy, room, for, one, conv...   
4  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descripcion_vecindad  cantidad_baños  \
0  Lower Queen Anne is near the Seattle Center (s...               1   
1  Upper Queen Anne is a really pleasant, unique ...               1   
2  The convenience of being in Seattle but on the...               1   
3  Ballard is lovely, vibrant and one of the most...               1   
4                                                 --               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
1                 1               1          Futon   
2                 1               1          Futon   
3                 1               1  Pull-out Sofa   
4                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
1  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
2  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
3  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
4  [Wireless Internet, Kitchen, Free Parking on P...             0.0   

   cuota_limpieza   precio  
0             0.0  11000.0  
1             0.0   4500.0  
2             0.0   5500.0  
3          2000.0   5200.0  
4          1500.0   8500.0

#### Tokenizacion descripcion_vecindad 

In [65]:
datos['descripcion_vecindad'][0]

"Lower Queen Anne is near the Seattle Center (space needle, EMP museum, Glass museum, Science Center and Children's museum). It's also near SAM sculpture park, stores, restaurants, SIFF theater and more."

In [67]:
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('[^a-zA-Z0-9\-\' ]',' ', regex=True)
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('(?<!\w)-(?!\w)',' ', regex=True)
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('[\(\)"]','',regex=True)

<>:1: SyntaxWarning: invalid escape sequence '\-'
<>:2: SyntaxWarning: invalid escape sequence '\w'
<>:3: SyntaxWarning: invalid escape sequence '\('
<>:1: SyntaxWarning: invalid escape sequence '\-'
<>:2: SyntaxWarning: invalid escape sequence '\w'
<>:3: SyntaxWarning: invalid escape sequence '\('
C:\Users\caray\AppData\Local\Temp\ipykernel_20204\2258382751.py:1: SyntaxWarning: invalid escape sequence '\-'
  datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('[^a-zA-Z0-9\-\' ]',' ', regex=True)
C:\Users\caray\AppData\Local\Temp\ipykernel_20204\2258382751.py:2: SyntaxWarning: invalid escape sequence '\w'
  datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('(?<!\w)-(?!\w)',' ', regex=True)
C:\Users\caray\AppData\Local\Temp\ipykernel_20204\2258382751.py:3: SyntaxWarning: invalid escape sequence '\('
  datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.replace('[\(\)"]','',regex=True)


In [68]:
datos['descripcion_vecindad'][0]

"Lower Queen Anne is near the Seattle Center  space needle  EMP museum  Glass museum  Science Center and Children's museum   It's also near SAM sculpture park  stores  restaurants  SIFF theater and more "

In [70]:
datos['descripcion_vecindad'] = datos['descripcion_vecindad'].str.split()

In [72]:
datos.head()

evaluacion_general experiencia_local  max_hospedes  \
0                10.0                --             1   
1                10.0                --             1   
2                10.0                --             1   
3                10.0                --             1   
4                10.0                --             1   

                                   descripcion_local  \
0  [this, clean, and, comfortable, one, bedroom, ...   
1  [our, century, old, upper, queen, anne, house,...   
2  [cozy, room, in, two-bedroom, apartment, along...   
3  [very, lovely, and, cozy, room, for, one, conv...   
4  [the, studio, at, mibbett, hollow', is, in, a,...   

                                descripcion_vecindad  cantidad_baños  \
0  [Lower, Queen, Anne, is, near, the, Seattle, C...               1   
1  [Upper, Queen, Anne, is, a, really, pleasant, ...               1   
2  [The, convenience, of, being, in, Seattle, but...               1   
3  [Ballard, is, lovely, vibrant, and, one, of, t...               1   
4                                                 []               1   

   cantidad_cuartos  cantidad_camas    modelo_cama  \
0                 1               1       Real Bed   
1                 1               1          Futon   
2                 1               1          Futon   
3                 1               1  Pull-out Sofa   
4                 1               1       Real Bed   

                                         comodidades  cuota_deposito  \
0  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
1  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
2  [TV, Internet, Wireless Internet, Kitchen, Fre...             0.0   
3  [Internet, Wireless Internet, Kitchen, Free Pa...             0.0   
4  [Wireless Internet, Kitchen, Free Parking on P...             0.0   

   cuota_limpieza   precio  
0             0.0  11000.0  
1             0.0   4500.0  
2             0.0   5500.0  
3          2000.0   5200.0  
4          1500.0   8500.0

## Datos de Tiempo


In [182]:
import pandas as pd
import numpy as np

In [183]:
df_datos = pd.read_json('inmuebles_disponibles.json')

In [184]:
df_datos.sample(10)

id       fecha  lugar_disponible   precio
172812  2242  2016-06-19              True  $250.00
339459   234  2016-01-13              True  $110.00
323175  2149  2016-06-02             False     None
205032  2376  2016-09-27              True   $86.00
216174  1739  2016-04-07              True  $275.00
110200  3202  2016-12-04              True  $160.00
60486    868  2016-09-21              True  $135.00
67508   2677  2016-12-17              True  $150.00
6069    2700  2016-08-20              True   $60.00
49578   1948  2016-11-02             False     None

In [185]:
df_datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                365000 non-null  int64 
 1   fecha             365000 non-null  object
 2   lugar_disponible  365000 non-null  bool  
 3   precio            270547 non-null  object
dtypes: bool(1), int64(1), object(2)
memory usage: 11.5+ MB


In [186]:
#pasamos a tipo datetime
df_datos['fecha'] = pd.to_datetime(df_datos['fecha'])

In [187]:
df_datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                365000 non-null  int64         
 1   fecha             365000 non-null  datetime64[ns]
 2   lugar_disponible  365000 non-null  bool          
 3   precio            270547 non-null  object        
dtypes: bool(1), datetime64[ns](1), int64(1), object(1)
memory usage: 11.5+ MB


In [188]:
df_datos.head()

id      fecha  lugar_disponible precio
0  857 2016-01-04             False   None
1  857 2016-01-05             False   None
2  857 2016-01-06             False   None
3  857 2016-01-07             False   None
4  857 2016-01-08             False   None

In [189]:
# Cuantos lugares estuvieron disponibles en el mes?
df_datos['fecha'].dt.strftime('%Y-%m')

0         2016-01
1         2016-01
2         2016-01
3         2016-01
4         2016-01
           ...   
364995    2016-12
364996    2016-12
364997    2016-12
364998    2017-01
364999    2017-01
Name: fecha, Length: 365000, dtype: object

In [190]:
subset = df_datos.groupby(df_datos['fecha'].dt.strftime('%Y-%m'))['lugar_disponible'].sum() #Esto solo funciona si la columna es booleana

In [191]:
subset

fecha
2016-01    16543
2016-02    20128
2016-03    23357
2016-04    22597
2016-05    23842
2016-06    23651
2016-07    22329
2016-08    22529
2016-09    22471
2016-10    23765
2016-11    23352
2016-12    24409
2017-01     1574
Name: lugar_disponible, dtype: int64

In [192]:
df_datos['precio']

0            None
1            None
2            None
3            None
4            None
           ...   
364995    $140.00
364996    $140.00
364997    $140.00
364998    $140.00
364999       None
Name: precio, Length: 365000, dtype: object

### Desafio

Como hay valores none, primero tratamos los datos como si fueran str, luego aplicamos los reemplazos y aplicamos el tipo flotante

In [193]:
df_datos['precio'].fillna('0.0', inplace = True)
df_datos['precio'] = df_datos['precio'].astype(str).replace(r'[$,]','',regex=True).astype(np.float64)


In [194]:
df_datos['precio']

0           0.0
1           0.0
2           0.0
3           0.0
4           0.0
          ...  
364995    140.0
364996    140.0
364997    140.0
364998    140.0
364999      0.0
Name: precio, Length: 365000, dtype: float64

In [195]:
df_datos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 365000 entries, 0 to 364999
Data columns (total 4 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                365000 non-null  int64         
 1   fecha             365000 non-null  datetime64[ns]
 2   lugar_disponible  365000 non-null  bool          
 3   precio            365000 non-null  float64       
dtypes: bool(1), datetime64[ns](1), float64(1), int64(1)
memory usage: 11.5 MB


# Proyecto Desafío 1: Ventas Online

El objetivo de este proyecto es analizar los resultados de un evento con los clientes de una empresa de venta online. Se recopiló un conjunto de datos que contiene los clientes que más gastaron en productos durante los 5 días de ventas, que es la duración del evento. Este análisis identificará al cliente con la mayor compra esta semana, quien recibirá un premio de la tienda, y posteriormente, puede ayudar a la empresa a crear nuevas estrategias para atraer más clientes.

La base de datos utilizada en este análisis es dados_vendas_clientes.json y contiene información importante sobre los clientes, como el nombre registrado del cliente, el monto total pagado al momento de la compra y el día de la compra.

Sabiendo esta información, el desafío del proyecto 1: ventas online será abrir la base de datos con Pandas y aplicar json_normalize.

In [73]:
import pandas as pd
import numpy as np

In [74]:
ruta = 'datos_desafio_1/dados_vendas_clientes.json'
ventas = pd.read_json(ruta)

In [75]:
ventas

dados_vendas
0  {'Data de venda': '06/06/2022', 'Cliente': ['@...
1  {'Data de venda': '07/06/2022', 'Cliente': ['I...
2  {'Data de venda': '08/06/2022', 'Cliente': ['I...
3  {'Data de venda': '09/06/2022', 'Cliente': ['J...
4  {'Data de venda': '10/06/2022', 'Cliente': ['M...

In [76]:
ventas = pd.json_normalize(ventas['dados_vendas'])

In [77]:
ventas

Data de venda                                            Cliente  \
0    06/06/2022  [@ANA _LUCIA 321, DieGO ARMANDIU 210, DieGO AR...   
1    07/06/2022  [Isabely JOanes 738, Isabely JOanes 738, Isabe...   
2    08/06/2022  [Isabely JOanes 738, JOãO Gabriel 671, Julya m...   
3    09/06/2022  [Julya meireles 914, MaRIA Julia 444, MaRIA Ju...   
4    10/06/2022  [MaRIA Julia 444, PEDRO PASCO 812, Paulo castr...   

                                Valor da compra  
0    [R$ 836,5, R$ 573,33, R$ 392,8, R$ 512,34]  
1  [R$ 825,31, R$ 168,07, R$ 339,18, R$ 314,69]  
2  [R$ 682,05, R$ 386,34, R$ 622,65, R$ 630,79]  
3   [R$ 390,3, R$ 759,16, R$ 334,47, R$ 678,78]  
4  [R$ 314,24, R$ 311,15, R$ 899,16, R$ 885,24]

In [78]:
clientes = ventas.explode('Cliente')

In [79]:
clientes

Data de venda             Cliente  \
0    06/06/2022     @ANA _LUCIA 321   
0    06/06/2022  DieGO ARMANDIU 210   
0    06/06/2022  DieGO ARMANDIU 210   
0    06/06/2022  DieGO ARMANDIU 210   
1    07/06/2022  Isabely JOanes 738   
1    07/06/2022  Isabely JOanes 738   
1    07/06/2022  Isabely JOanes 738   
1    07/06/2022  Isabely JOanes 738   
2    08/06/2022  Isabely JOanes 738   
2    08/06/2022    JOãO Gabriel 671   
2    08/06/2022  Julya meireles 914   
2    08/06/2022  Julya meireles 914   
3    09/06/2022  Julya meireles 914   
3    09/06/2022     MaRIA Julia 444   
3    09/06/2022     MaRIA Julia 444   
3    09/06/2022     MaRIA Julia 444   
4    10/06/2022     MaRIA Julia 444   
4    10/06/2022     PEDRO PASCO 812   
4    10/06/2022    Paulo castro 481   
4    10/06/2022   Thiago fritzz 883   

                                Valor da compra  
0    [R$ 836,5, R$ 573,33, R$ 392,8, R$ 512,34]  
0    [R$ 836,5, R$ 573,33, R$ 392,8, R$ 512,34]  
0    [R$ 836,5, R$ 573,33, R$ 392,8, R$ 512,34]  
0    [R$ 836,5, R$ 573,33, R$ 392,8, R$ 512,34]  
1  [R$ 825,31, R$ 168,07, R$ 339,18, R$ 314,69]  
1  [R$ 825,31, R$ 168,07, R$ 339,18, R$ 314,69]  
1  [R$ 825,31, R$ 168,07, R$ 339,18, R$ 314,69]  
1  [R$ 825,31, R$ 168,07, R$ 339,18, R$ 314,69]  
2  [R$ 682,05, R$ 386,34, R$ 622,65, R$ 630,79]  
2  [R$ 682,05, R$ 386,34, R$ 622,65, R$ 630,79]  
2  [R$ 682,05, R$ 386,34, R$ 622,65, R$ 630,79]  
2  [R$ 682,05, R$ 386,34, R$ 622,65, R$ 630,79]  
3   [R$ 390,3, R$ 759,16, R$ 334,47, R$ 678,78]  
3   [R$ 390,3, R$ 759,16, R$ 334,47, R$ 678,78]  
3   [R$ 390,3, R$ 759,16, R$ 334,47, R$ 678,78]  
3   [R$ 390,3, R$ 759,16, R$ 334,47, R$ 678,78]  
4  [R$ 314,24, R$ 311,15, R$ 899,16, R$ 885,24]  
4  [R$ 314,24, R$ 311,15, R$ 899,16, R$ 885,24]  
4  [R$ 314,24, R$ 311,15, R$ 899,16, R$ 885,24]  
4  [R$ 314,24, R$ 311,15, R$ 899,16, R$ 885,24]

In [80]:
clientes = clientes.explode('Valor da compra')


In [81]:
clientes.head()

Data de venda             Cliente Valor da compra
0    06/06/2022     @ANA _LUCIA 321        R$ 836,5
0    06/06/2022     @ANA _LUCIA 321       R$ 573,33
0    06/06/2022     @ANA _LUCIA 321        R$ 392,8
0    06/06/2022     @ANA _LUCIA 321       R$ 512,34
0    06/06/2022  DieGO ARMANDIU 210        R$ 836,5

In [82]:
#Transformamos columnas tipo object que se ve que son numericas, a tipo float
clientes['Valor da compra'] = clientes['Valor da compra'].apply(lambda x: x.replace('R$','').replace(',','.').strip()).astype(np.float64)

In [83]:
clientes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 80 entries, 0 to 4
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Data de venda    80 non-null     object 
 1   Cliente          80 non-null     object 
 2   Valor da compra  80 non-null     float64
dtypes: float64(1), object(2)
memory usage: 2.5+ KB


In [84]:
clientes.head()

Data de venda             Cliente  Valor da compra
0    06/06/2022     @ANA _LUCIA 321           836.50
0    06/06/2022     @ANA _LUCIA 321           573.33
0    06/06/2022     @ANA _LUCIA 321           392.80
0    06/06/2022     @ANA _LUCIA 321           512.34
0    06/06/2022  DieGO ARMANDIU 210           836.50

#### Parte 3 Proyecto 1

In [ ]:
clientes['Cliente'] = clientes['Cliente'].str.replace('[^a-zA-Z]','', regex=True)

In [87]:
clientes['Cliente'] = clientes['Cliente'].str.lower()

In [88]:
clientes.head()

Data de venda        Cliente  Valor da compra
0    06/06/2022       analucia           836.50
0    06/06/2022       analucia           573.33
0    06/06/2022       analucia           392.80
0    06/06/2022       analucia           512.34
0    06/06/2022  diegoarmandiu           836.50

In [196]:
clientes['Data de venda'] = pd.to_datetime(clientes['Data de venda'])

In [204]:
clientes

Data de venda        Cliente  Valor da compra
0     2022-06-06       analucia           836.50
0     2022-06-06       analucia           573.33
0     2022-06-06       analucia           392.80
0     2022-06-06       analucia           512.34
0     2022-06-06  diegoarmandiu           836.50
..           ...            ...              ...
4     2022-10-06    paulocastro           885.24
4     2022-10-06   thiagofritzz           314.24
4     2022-10-06   thiagofritzz           311.15
4     2022-10-06   thiagofritzz           899.16
4     2022-10-06   thiagofritzz           885.24

[80 rows x 3 columns]

In [223]:
subset = clientes.groupby(clientes['Cliente'])['Valor da compra'].sum()

AttributeError: 'Series' object has no attribute 'sort'

In [213]:
subset

Cliente
analucia         2314.97
diegoarmandiu    6944.91
isabelyjoanes    8910.83
joogabriel       2321.83
julyameireles    6806.37
mariajulia       8897.92
paulocastro      2409.79
pedropasco       2409.79
thiagofritzz     2409.79
Name: Valor da compra, dtype: float64

### Ahora seleccionamos al que gasto mas, podemos simplemente aplicar sort_values(ascending=False) y tomar al primero

In [233]:
subset = subset.sort_values(ascending=False)

In [235]:
subset

Cliente
isabelyjoanes    8910.83
mariajulia       8897.92
diegoarmandiu    6944.91
julyameireles    6806.37
paulocastro      2409.79
pedropasco       2409.79
thiagofritzz     2409.79
joogabriel       2321.83
analucia         2314.97
Name: Valor da compra, dtype: float64

In [239]:
print(f'El cliente premiado es: {subset.index[0]}')

El cliente premiado es: isabelyjoanes


### Opcion con idxmax()
* devuelve la etiqueta del índice correspondiente al valor máximo de una Serie o columna de un DataFrame.
* Ej
    * 'Ana': 150,
    * 'Luis': 320,
    * 'Carlos': 280
* Resultado = Luis 

In [241]:
print(f'El cliente premiado es: {subset.idxmax()}')

El cliente premiado es: isabelyjoanes


# Proyecto Desafío 2: Administración de Condominios

In [89]:
import pandas as pd

In [90]:
condominios = pd.read_json('datos_desafio_2/dados_locacao_imoveis.json')
condominios.head()

dados_locacao
0  {'apartamento': 'A101 (blocoAP)', 'datas_combi...
1  {'apartamento': 'A102 (blocoAP)', 'datas_combi...
2  {'apartamento': 'B201 (blocoAP)', 'datas_combi...
3  {'apartamento': 'B202 (blocoAP)', 'datas_combi...
4  {'apartamento': 'C301 (blocoAP)', 'datas_combi...

In [91]:
condominios = pd.json_normalize(condominios['dados_locacao'])
condominios

apartamento datas_combinadas_pagamento        datas_de_pagamento  \
0   A101 (blocoAP)   [01/06/2022, 01/07/2022]  [05/06/2022, 03/07/2022]   
1   A102 (blocoAP)   [02/06/2022, 02/07/2022]  [02/06/2022, 06/07/2022]   
2   B201 (blocoAP)   [03/06/2022, 03/07/2022]  [07/06/2022, 03/07/2022]   
3   B202 (blocoAP)   [04/06/2022, 04/07/2022]  [07/06/2022, 05/07/2022]   
4   C301 (blocoAP)   [05/06/2022, 05/07/2022]  [10/06/2022, 09/07/2022]   
5   C302 (blocoAP)   [06/06/2022, 06/07/2022]  [08/06/2022, 12/07/2022]   
6   D401 (blocoAP)   [07/06/2022, 07/07/2022]  [07/06/2022, 09/07/2022]   
7   D402 (blocoAP)   [08/06/2022, 08/07/2022]  [10/06/2022, 14/07/2022]   
8   E501 (blocoAP)   [09/06/2022, 09/07/2022]  [10/06/2022, 09/07/2022]   
9   E502 (blocoAP)   [10/06/2022, 10/07/2022]  [16/06/2022, 12/07/2022]   
10  F601 (blocoAP)   [11/06/2022, 11/07/2022]  [17/06/2022, 13/07/2022]   
11  F602 (blocoAP)   [12/06/2022, 12/07/2022]  [15/06/2022, 12/07/2022]   
12  G701 (blocoAP)   [13/06/2022, 13/07/2022]  [19/06/2022, 20/07/2022]   
13  G702 (blocoAP)   [14/06/2022, 14/07/2022]  [15/06/2022, 17/07/2022]   
14  H801 (blocoAP)   [15/06/2022, 15/07/2022]  [18/06/2022, 16/07/2022]   

                       valor_aluguel  
0   [$ 1000,0 reais, $ 2500,0 reais]  
1   [$ 1100,0 reais, $ 2600,0 reais]  
2   [$ 1200,0 reais, $ 2700,0 reais]  
3   [$ 1300,0 reais, $ 2800,0 reais]  
4   [$ 1400,0 reais, $ 2900,0 reais]  
5   [$ 1500,0 reais, $ 1200,0 reais]  
6   [$ 1600,0 reais, $ 1300,0 reais]  
7   [$ 1700,0 reais, $ 1400,0 reais]  
8   [$ 1800,0 reais, $ 1500,0 reais]  
9   [$ 1900,0 reais, $ 1600,0 reais]  
10  [$ 2000,0 reais, $ 1000,0 reais]  
11  [$ 2100,0 reais, $ 1100,0 reais]  
12  [$ 2200,0 reais, $ 1200,0 reais]  
13  [$ 2300,0 reais, $ 2100,0 reais]  
14  [$ 2400,0 reais, $ 2200,0 reais]

### Convirtiendo a datatime las columnas con fechas

In [92]:
condominios['datas_combinadas_pagamento'] = condominios['datas_combinadas_pagamento'].apply(
    lambda lista: [pd.to_datetime(fecha, dayfirst=True).date() for fecha in lista]
)

In [93]:
condominios['datas_de_pagamento'] = condominios['datas_de_pagamento'].apply(
    lambda lista: [pd.to_datetime(fecha, dayfirst=True).date() for fecha in lista]
)

### Calculando dias de retraso

Esta funcion: 
* recorre las listas **fechas_pagadas y fechas_acordadas** al mismo tiempo
* **(pago - acordado).days** extrae los dias de retraso

In [94]:
# Calcular retrasos en días
def calcular_retrasos(fechas_acordadas, fechas_pagadas):
    return [(pago - acordado).days for pago, acordado in zip(fechas_pagadas, fechas_acordadas)]

Se crea una nueva columna llamada **dias_retraso**, donde se aplica la funcion anterior fila por fila.
* Ejemplo: 
    * Fechas acordadas para pagar **[01/06/2022, 01/07/2022]**
    * Fechas donde se pago **[05/06/2022, 03/07/2022]**
* Resultado: **[4,2]**, ya que de **01/06/2022 al 05/06/2022** se tiene 4 dias de retraso, y de **01/07/2022 a 03/07/2022** son 2 dias de retraso

In [95]:
condominios['dias_retraso'] = condominios.apply(
    lambda row: calcular_retrasos(row['datas_combinadas_pagamento'], row['datas_de_pagamento']),
    axis=1
)

In [96]:
# Indicar si hubo retraso
condominios['pago_con_retraso'] = condominios['dias_retraso'].apply(
    lambda lista: [dias > 0 for dias in lista]
)

In [97]:
condominios.head()

apartamento datas_combinadas_pagamento        datas_de_pagamento  \
0  A101 (blocoAP)   [2022-06-01, 2022-07-01]  [2022-06-05, 2022-07-03]   
1  A102 (blocoAP)   [2022-06-02, 2022-07-02]  [2022-06-02, 2022-07-06]   
2  B201 (blocoAP)   [2022-06-03, 2022-07-03]  [2022-06-07, 2022-07-03]   
3  B202 (blocoAP)   [2022-06-04, 2022-07-04]  [2022-06-07, 2022-07-05]   
4  C301 (blocoAP)   [2022-06-05, 2022-07-05]  [2022-06-10, 2022-07-09]   

                      valor_aluguel dias_retraso pago_con_retraso  
0  [$ 1000,0 reais, $ 2500,0 reais]       [4, 2]     [True, True]  
1  [$ 1100,0 reais, $ 2600,0 reais]       [0, 4]    [False, True]  
2  [$ 1200,0 reais, $ 2700,0 reais]       [4, 0]    [True, False]  
3  [$ 1300,0 reais, $ 2800,0 reais]       [3, 1]     [True, True]  
4  [$ 1400,0 reais, $ 2900,0 reais]       [5, 4]     [True, True]

In [98]:
condominios = condominios.explode(['datas_combinadas_pagamento','datas_de_pagamento','valor_aluguel','dias_retraso', 'pago_con_retraso'])

In [99]:
condominios['valor_aluguel'] = condominios['valor_aluguel'].apply(lambda x: x.replace('reais','').replace(',','.').replace('$','').strip()).astype(np.float64)

In [100]:
condominios

apartamento datas_combinadas_pagamento datas_de_pagamento  \
0   A101 (blocoAP)                 2022-06-01         2022-06-05   
0   A101 (blocoAP)                 2022-07-01         2022-07-03   
1   A102 (blocoAP)                 2022-06-02         2022-06-02   
1   A102 (blocoAP)                 2022-07-02         2022-07-06   
2   B201 (blocoAP)                 2022-06-03         2022-06-07   
2   B201 (blocoAP)                 2022-07-03         2022-07-03   
3   B202 (blocoAP)                 2022-06-04         2022-06-07   
3   B202 (blocoAP)                 2022-07-04         2022-07-05   
4   C301 (blocoAP)                 2022-06-05         2022-06-10   
4   C301 (blocoAP)                 2022-07-05         2022-07-09   
5   C302 (blocoAP)                 2022-06-06         2022-06-08   
5   C302 (blocoAP)                 2022-07-06         2022-07-12   
6   D401 (blocoAP)                 2022-06-07         2022-06-07   
6   D401 (blocoAP)                 2022-07-07         2022-07-09   
7   D402 (blocoAP)                 2022-06-08         2022-06-10   
7   D402 (blocoAP)                 2022-07-08         2022-07-14   
8   E501 (blocoAP)                 2022-06-09         2022-06-10   
8   E501 (blocoAP)                 2022-07-09         2022-07-09   
9   E502 (blocoAP)                 2022-06-10         2022-06-16   
9   E502 (blocoAP)                 2022-07-10         2022-07-12   
10  F601 (blocoAP)                 2022-06-11         2022-06-17   
10  F601 (blocoAP)                 2022-07-11         2022-07-13   
11  F602 (blocoAP)                 2022-06-12         2022-06-15   
11  F602 (blocoAP)                 2022-07-12         2022-07-12   
12  G701 (blocoAP)                 2022-06-13         2022-06-19   
12  G701 (blocoAP)                 2022-07-13         2022-07-20   
13  G702 (blocoAP)                 2022-06-14         2022-06-15   
13  G702 (blocoAP)                 2022-07-14         2022-07-17   
14  H801 (blocoAP)                 2022-06-15         2022-06-18   
14  H801 (blocoAP)                 2022-07-15         2022-07-16   

    valor_aluguel dias_retraso pago_con_retraso  
0          1000.0            4             True  
0          2500.0            2             True  
1          1100.0            0            False  
1          2600.0            4             True  
2          1200.0            4             True  
2          2700.0            0            False  
3          1300.0            3             True  
3          2800.0            1             True  
4          1400.0            5             True  
4          2900.0            4             True  
5          1500.0            2             True  
5          1200.0            6             True  
6          1600.0            0            False  
6          1300.0            2             True  
7          1700.0            2             True  
7          1400.0            6             True  
8          1800.0            1             True  
8          1500.0            0            False  
9          1900.0            6             True  
9          1600.0            2             True  
10         2000.0            6             True  
10         1000.0            2             True  
11         2100.0            3             True  
11         1100.0            0            False  
12         2200.0            6             True  
12         1200.0            7             True  
13         2300.0            1             True  
13         2100.0            3             True  
14         2400.0            3             True  
14         2200.0            1             True

#### Parte 3, elimminar (blocoAP)

In [103]:
condominios['apartamento'] = condominios['apartamento'].str.replace('[blocoAP\(\)]','',regex=True)

<>:1: SyntaxWarning: invalid escape sequence '\('
<>:1: SyntaxWarning: invalid escape sequence '\('
C:\Users\caray\AppData\Local\Temp\ipykernel_20204\1860703611.py:1: SyntaxWarning: invalid escape sequence '\('
  condominios['apartamento'] = condominios['apartamento'].str.replace('[blocoAP\(\)]','',regex=True)


In [104]:
condominios

apartamento datas_combinadas_pagamento datas_de_pagamento  valor_aluguel  \
0         101                  2022-06-01         2022-06-05         1000.0   
0         101                  2022-07-01         2022-07-03         2500.0   
1         102                  2022-06-02         2022-06-02         1100.0   
1         102                  2022-07-02         2022-07-06         2600.0   
2        B201                  2022-06-03         2022-06-07         1200.0   
2        B201                  2022-07-03         2022-07-03         2700.0   
3        B202                  2022-06-04         2022-06-07         1300.0   
3        B202                  2022-07-04         2022-07-05         2800.0   
4        C301                  2022-06-05         2022-06-10         1400.0   
4        C301                  2022-07-05         2022-07-09         2900.0   
5        C302                  2022-06-06         2022-06-08         1500.0   
5        C302                  2022-07-06         2022-07-12         1200.0   
6        D401                  2022-06-07         2022-06-07         1600.0   
6        D401                  2022-07-07         2022-07-09         1300.0   
7        D402                  2022-06-08         2022-06-10         1700.0   
7        D402                  2022-07-08         2022-07-14         1400.0   
8        E501                  2022-06-09         2022-06-10         1800.0   
8        E501                  2022-07-09         2022-07-09         1500.0   
9        E502                  2022-06-10         2022-06-16         1900.0   
9        E502                  2022-07-10         2022-07-12         1600.0   
10       F601                  2022-06-11         2022-06-17         2000.0   
10       F601                  2022-07-11         2022-07-13         1000.0   
11       F602                  2022-06-12         2022-06-15         2100.0   
11       F602                  2022-07-12         2022-07-12         1100.0   
12       G701                  2022-06-13         2022-06-19         2200.0   
12       G701                  2022-07-13         2022-07-20         1200.0   
13       G702                  2022-06-14         2022-06-15         2300.0   
13       G702                  2022-07-14         2022-07-17         2100.0   
14       H801                  2022-06-15         2022-06-18         2400.0   
14       H801                  2022-07-15         2022-07-16         2200.0   

   dias_retraso pago_con_retraso  
0             4             True  
0             2             True  
1             0            False  
1             4             True  
2             4             True  
2             0            False  
3             3             True  
3             1             True  
4             5             True  
4             4             True  
5             2             True  
5             6             True  
6             0            False  
6             2             True  
7             2             True  
7             6             True  
8             1             True  
8             0            False  
9             6             True  
9             2             True  
10            6             True  
10            2             True  
11            3             True  
11            0            False  
12            6             True  
12            7             True  
13            1             True  
13            3             True  
14            3             True  
14            1             True

In [242]:
media_atraso = condominios.groupby(['apartamento'])['dias_retraso'].mean()

In [243]:
media_atraso

apartamento
101      3.0
102      2.0
B201     2.0
B202     2.0
C301     4.5
C302     4.0
D401     1.0
D402     4.0
E501     0.5
E502     4.0
F601     4.0
F602     1.5
G701     6.5
G702     2.0
H801     2.0
Name: dias_retraso, dtype: object